### PR TITLE
feat: implement custom site-wide roles in the database

### PR DESCRIPTION
**Pull request by Emyrk**
# What this does

Adds the ability to create custom roles at the site level. Intentionally not doing org scoped roles yet, as org roles cannot be seen on the UI yet.

This includes the api endpoint to create a custom role. Ideally that would be another PR, but I wanted to actually test the custom roles through the api actor logic. Roles and the actor are so intertwined, if the custom roles do not work to actually grant perms at the api layer, then the feature is moot.

# Future work

- Cli interface
- UI interface
- Org roles (not for awhile)
- Fetch roles api needs to include permissions
- Delete roles

### CODE REVIEW

**Comment by Emyrk** [comment]
* **#13290** <a href="https://app.graphite.dev/github/pr/coder/coder/13290?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#13289** <a href="https://app.graphite.dev/github/pr/coder/coder/13289?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Emyrk and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: coderd/database/db2sdk/db2sdk.go [hunk 1]
```diff
@@ -28,9 +29,25 @@ import (
...
+	}
+}
+
+func Map[K comparable, F any, T any](params map[K]F, convert func(F) T) map[K]T {
+	into := make(map[K]T)
+	for k, item := range params {
+		into[k] = convert(item)
 	}
 	return into
 }
```
Comment (johnstcn) [inline coderd/database/db2sdk/db2sdk.go:53]: All of these should probably be in `coderd/util`

File: codersdk/roles.go [hunk 2]
```diff
@@ -9,16 +9,48 @@ import (
...
+type SlimRole struct {
 	Name        string `json:"name"`
 	DisplayName string `json:"display_name"`
 }
 
 type AssignableRoles struct {
-	Role
+	SlimRole
 	Assignable bool `json:"assignable"`
 }
```
Comment (johnstcn) [inline codersdk/roles.go:20]: Can you provide a comment explaining the difference between `Role` and `SlimRole`?

File: enterprise/coderd/roles_test.go
Comment (johnstcn) [inline enterprise/coderd/roles_test.go:1]: I'd like to see more tests:
- What happens if we have existing custom roles and our license expires? 
- What happens if we patch a role and leave out certain fields in the request? Do the unspecified values get overwritten in the DB?

**Review by johnstcn — COMMENTED** [review]
This PR is a bit too large for me to review comforably. 
Can you split it up a bit?
Here's my suggestion, but feel free to modify:
- Introduce new DB types / queries etc.
- Add required dbauthz changes
- Plumb in new HTTP middleware etc.

**Comment by Emyrk** [comment]
Split into a stack: https://github.com/coder/coder/pull/13298

**Pull request closed without merging**